validator/project_validator_test.go

```diff
@@ -21,6 +27,94 @@ import (
 	"github.com/stretchr/testify/suite"
 )
 
+func TestProjectErrorValidators(t *testing.T) {
+	node, err := parser.ParseFile(token.NewFileSet(), "project_validator.go", nil, parser.AllErrors)
+	require.NoError(t, err)
+	funcBodies := make(map[string]*ast.BlockStmt)
+	ast.Inspect(node, func(n ast.Node) bool {
+		if fn, ok := n.(*ast.FuncDecl); ok {
+			funcBodies[fn.Name.Name] = fn.Body
+		}
+		return true
+	})
+
+	// projectErrorValidators have some restrictions and conventions that they must follow:
+	// 1. They must return an error explicitly.
+	// 2. They must not return any other type of ValidationError level.
+	for _, validator := range projectErrorValidators {
+		funcPtr := runtime.FuncForPC(reflect.ValueOf(validator).Pointer())
+		funcName := funcPtr.Name()[strings.LastIndex(funcPtr.Name(), ".")+1:]
+
+		t.Run(funcName, func(t *testing.T) {
+			assert.True(t, variablesInFunction(funcBodies, funcName, []string{"Error"}, map[string]bool{}), "ProjectErrorValidators should return at least one Error")
+			assert.False(t, variablesInFunction(funcBodies, funcName, []string{"Warning", "Notice"}, map[string]bool{}), "ProjectErrorValidators should never use Warnings or Notices")
+		})
+	}
+}
+
+func TestProjectWarningValidators(t *testing.T) {
+	node, err := parser.ParseFile(token.NewFileSet(), "project_validator.go", nil, parser.AllErrors)
+	require.NoError(t, err)
+	funcBodies := make(map[string]*ast.BlockStmt)
+	ast.Inspect(node, func(n ast.Node) bool {
+		if fn, ok := n.(*ast.FuncDecl); ok {
+			funcBodies[fn.Name.Name] = fn.Body
+		}
+		return true
+	})
+
+	// projectWarningValidators must only return Warning or Notice.
+	for _, validator := range projectWarningValidators {
+		funcPtr := runtime.FuncForPC(reflect.ValueOf(validator).Pointer())
+		funcName := funcPtr.Name()[strings.LastIndex(funcPtr.Name(), ".")+1:]
+
+		t.Run(funcName, func(t *testing.T) {
+			assert.False(t, variablesInFunction(funcBodies, funcName, []string{"Error"}, map[string]bool{}), "ProjectWarningValidators should never use Error")
+			assert.True(t, variablesInFunction(funcBodies, funcName, []string{"Warning", "Notice"}, map[string]bool{}), "ProjectWarningValidators return at least one Warning or Notice")
+		})
+	}
+}
+
+// variablesInFunction recursively checks if the given variables are used in a function or any of the functions it calls.
+func variablesInFunction(funcBodies map[string]*ast.BlockStmt, funcName string, variables []string, visited map[string]bool) bool {
```

Comment:
nit: since this is doing reflection, might be clearer to rename `variables` to `variableNames` to indicate we're literally looking for the enum's variable names.